system_modules/napsequence/src/sequenceplayer.cpp

```diff
@@ -218,6 +219,58 @@ namespace nap
     }
 
 
+    bool SequencePlayer::loadBinary(const std::vector<uint8>& buffer, utility::ErrorState& errorState)
+    {
+        std::lock_guard<std::mutex> lock(mMutex);
+        rtti::DeserializeResult result;
+
+        //
+        auto stream = utility::MemoryStream(buffer.data(), buffer.size());
+        rtti::Factory factory;
+        if(!rtti::deserializeBinary(stream, factory, result, errorState))
```

Comment:
Are you deserializing using raw pointers? Threaded deserialization without using raw pointers is not thread safe. From `epropertyvalidationmode.h`:

```
		/**
		 * The deserializer can operate in two modes: using only raw pointers or without any raw pointers at all. The main distinction
		 * between the two modes lie in hotloading: in the normal flow of operation, all objects are managed by ResourceManager and
		 * internally by ObjectPtrManager. These classes enable the hotloading of objects and they do so by patching internal pointers 
		 * in non-raw pointers like ObjectPtr, ResourcePtr, ComponentPtr. In such situations, raw pointers are not allowed. 
		 * However, ObjectPtr and alike are not thread-safe and they should only be used from the main thread. In cases where objects
		 * are deserialized directly without the use of ResourceManager, you can use raw pointers instead of ObjectPtr and alike. 
		 * By enabling OnlyRawPointers or NoRawPointers, the deserializer verifies the correct use of all pointers throughout deserialization.
		 */
```

Reply:
No, SequenceTrack and it's owned Segments are all rtti::Objects.

Rewriting this would involve a complete overhaul, something I would not like to venture into....

Sequences are always serialized / deserialized by the SequencePlayer and the sequence player is responsible for thread-safety regarding this

Reply:
Is the SequencePlayer saving / loading from the main thread? In that case it is _acceptable_. But why would this involve a complete overhaul? If you deserialize using raw pointers you're simply not involving the resource manager, something you're not interested in (right?) because everything is handled internal by the SequencePlayer? Why does the resource manager have to be aware of your objects? 

Deserializatrion using raw pointers was added because I was running into serious race-conditions deserializing api messages on a separate thread -> causing sporadic hangs if a pointer was patched the moment a message was deserialized. Instead of introducing a mutex we added raw pointer support. The only thing required from the developer is that you delete the objects yourself and that the objects are not patched, which isn't a requirement here anyway?

What's happing now is that you add many objects to the resource manager that will never be patched because the management is handled by you and not the init (reload) system.

Something to consider.

Reply:
The JSON sequences are written and loaded from the main thread, but since all SequenceSegments and SequenceTracks inherit from rtti::ObjectPtrs (also uses math::FCurve which inherits from ResourcePtr) I don't see how you can serialize/deserialize a sequence JSON only using raw pointers.

Also, I don't see an option for `deserializeBinary` to allow or not allow raw pointers...

So this leaves me a bit puzzled....

Reply:
> So this leaves me a bit puzzled....

You _can_ have a struct that has resource pointers, see `MaterialInstanceResource`. But it's not the best practice, although supported. 

> Also, I don't see an option for deserializeBinary to allow or not allow raw pointers...

That's something I can look into. As long as you use the main thread to serialize /  deserialize your sequences everything should be ok. For future reference, if you can embed something local and it's a light weight object, consider declaring it as a `STRUCT`

Reply:
ok, I would like to have a look at how the data of sequences is stored and get rid of all the tracks having pointers to segments but have everything nicely embedded and readable.

 I agree I should have worked with STRUCTS from the get-go, however I would like to do this for a major upgrade because this will also break older sequences and would like to make a PR for this solely addressing this issue.

Reply:
> ok, I would like to have a look at how the data of sequences is stored and get rid of all the tracks having pointers to segments and have everything nicely embedded and readable.
> 
> I agree I should have worked with STRUCTS from the get-go, however I would like to do this for a major new release because this will also break older sequences and would like to make a PR for this solely addressing this issue.

Yeah, if it's a breaking change it will have to wait for 0.8, which is the next major release.